rpcs3/Emu/SysCalls/Modules/cellPngDec.cpp

```diff
@@ -3,572 +3,859 @@
+	handle->malloc = thread_in_param->cbCtrlMallocFunc;
+	handle->malloc_arg = thread_in_param->cbCtrlMallocArg;
+	handle->free = thread_in_param->cbCtrlFreeFunc;
+	handle->free_arg = thread_in_param->cbCtrlFreeArg;
+
+	// Set handle pointer
+	*png_handle = handle;
+
+	// Set the version information
+	thread_out_param->pngCodecVersion = PNGDEC_CODEC_VERSION;
+
 	return CELL_OK;
 }
 
-s32 pngReadHeader(PSubHandle stream, vm::ptr<CellPngDecInfo> info, PExtInfo extInfo = vm::null)
+s32 pngDecDestroy(PPUThread& ppu, PHandle handle)
 {
-	CellPngDecInfo& current_info = stream->info;
+	// Deallocate the decoder handle memory
+	if (handle->free(ppu, handle, handle->free_arg) != 0)
+	{
+		cellPngDec.error("PNG decoder deallocation failed.");
+		return CELL_PNGDEC_ERROR_FATAL;
+	}
+
+	return CELL_OK;
+}
 
-	if (stream->fileSize < 29)
+s32 pngDecOpen(PPUThread& ppu, PHandle handle, PPStream png_stream, PSrc source, POpenInfo open_info, PCbControlStream control_stream = vm::null, POpenParam open_param = vm::null)
+{
+	// Check if partial image decoding is used
+	if (control_stream || open_param)
 	{
-		return CELL_PNGDEC_ERROR_HEADER; // The file is smaller than the length of a PNG header
+		throw EXCEPTION("Partial image decoding is not supported.");
 	}
 
-	// Write the header to buffer
-	u8 buffer[34]; be_t<u32>* buffer_32 = reinterpret_cast<be_t<u32>*>(buffer);
+	// Allocate memory for the stream structure
+	auto stream = PStream::make(handle->malloc(ppu, sizeof(PngStream), handle->malloc_arg).addr());
 
-	switch (stream->src.srcSelect)
+	// Check if the allocation of memory for the stream structure failed
+	if (!stream)
 	{
-	case CELL_PNGDEC_BUFFER:
-		std::memcpy(buffer, stream->src.streamPtr.get_ptr(), sizeof(buffer));
-		break;
-	case CELL_PNGDEC_FILE:
+		cellPngDec.error("PNG stream creation failed.");
+		return CELL_PNGDEC_ERROR_FATAL;
+	}
+
+	// Set memory info
+	open_info->initSpaceAllocated = sizeof(PngStream);
+
+	// Set the stream source to the source give by the game
+	stream->source = *source;
+
+	// Indicate that a fixed alpha value isn't used, if not specified otherwise
+	stream->fixed_alpha = false;
+
+	// Use virtual memory address as a handle
+	*png_stream = stream;
+
+	// Make sure the file pointer is a nullptr, before trying to open a file
+	stream->file_pointer = nullptr;
+
+	// Make sure the PngBuffer pointer is a nullptr, so we don't try to free it later
+	stream->buffer = vm::ptr<PngBuffer>::make((uintptr_t)nullptr);
+
+	// Open the buffer/file and check the header
+	u8 header[8];
+
+	// Depending on the source type, get the first 8 bytes
+	if (source->srcSelect == CELL_PNGDEC_FILE)
 	{
-		auto file = idm::get<lv2_file_t>(stream->fd);
-		file->file->Seek(0);
-		file->file->Read(buffer, sizeof(buffer));
-		break;
+		// Get the absolute path and open the file
+		std::string real_path;
+		Emu.GetVFS().GetDevice(stream->source.fileName.get_ptr(), real_path);
+		stream->file_pointer = fopen(real_path.c_str(), "rb");
+
```

Comment:
Why you replaced original lv2_file_t implementation? It wasn't generally wrong, and placing real pointers (or real handles) in PS3 memory is wrong.